lib/awful/tag.lua

```diff
@@ -718,6 +719,22 @@ end
 -- @tparam layout|function layout A layout table or a constructor function
 -- @return The layout
 
+--- The (proposed) list of available layouts for this tag.
+--
+-- This property allows to define a subset (or superset) of layouts available
+-- in the "rotation table". In the default configuration file, `Mod4+Space`
+-- and `Mod4+Shift+Space` are used to switch between tags. The
+-- `awful.widget.layoutlist` also uses this as its default layout filter.
+--
+-- By default, it will be the same as `awful.layout.layouts` unless there the
+-- a layout not present is used. If that's the case they will be added at the
```

Comment:
"there the a"?

How about:

> By default, it will be the same as `awful.layout.layouts` unless the tag currently uses a layout that is not listed there. In this case, the current layout will be prepended to the front of the list.

And why this prepending behaviour? Currently we also don't auto-magically add things to the list of layouts.

Reply:
> Currently we also don't auto-magically add things to the list of layouts.

You sure can:

```
awful.layout.layout = {awful.layout.suit.max}

awful.tag.add("mytag", {
    layout = awful.layout.suit.tile.left,
    layouts = {awful.layout.suit.tile.right},
})
```

With things like the lain layouts, this will totally happen. Plus, if you got a "global" `layoutlist` widget in your wibar and click on a layout not part of `mytag.layouts`, then it will trigger the error/bug too when you display the `layoutlist` for this tag. Instead of trying to fix the error in the `layoutlist`, better handle it at the source, here.

Reply:
Huh? How does the order of layouts fix bugs in a layoutlist widget? Why is the same bug not triggered if I just toggle through the layouts until I am at the end of the layout list?
Put differently: A layoutlist widget only cares about the current layout (or the list of available layouts, if it presents all of them in some way). I don't see how the order is relevant here.

Reply:
There is some cases:

 * You use `awful.layout.layouts` like the default config. Then there is no problems.
 * You use `awful.layout.layouts` but have code somehow setting layouts not part of it. Then the layoutlist and `mod4+space` based on it can't rotate anymore because the current layout is not part of the list.
 * You use `t.layouts` and your current tag layout is part of it: there is no problem.
 * You use `t.layouts` but have a different `t.layout`: see above
 * You have `t.layouts` but also have a `awful.widget.layoutlist` which uses `awful.layout.layouts`. Then the list mismatches and again you can't rotate properly unless you keep adding each layout from `awful.layout.layouts` when you rotate.
 * You use layouts objects, then you lose the state when rotating, but it wont happen because `awful.widget.layoutlist` holds a reference so it wont get deleted until the popup/widget is gone. So `t.layouts` will then be able to GC itself later when the state isn't "locked" anymore.

> Huh? What would it leak? If it leaks a layout, then I still have to ask again: "why is it okay when the list of available layouts on a tag changes randomly" (where "randomly" depends on the behaviour of the GC)

Assuming a bunch of things, including it relates closely to the `layoutlist` not GC-ing things currently displayed, then I think it's close enough to "be ok". I mean, `t.layouts` wont change while you use it because the result will "lock" the objects with non-weak references. There is the case that if you close the layoutlist then open it again very fast it will still have "too many things" compared to waiting a couple seconds. However is it really a problem in practice?

As for "what if you have a global list in the wibar and rotate it, then the temporary entries wont get removed" corner case, the `layoutlist` PR adds a `:cleanup()` function you can connect to the tag change signal.

> How does the order of layouts fix bugs in a layoutlist widget?

* Rotating the list is important (that's what the default config does), the corrent layout need to be part of that rotation even if it isn't in the list.
* It looks weird if there is no selected item, which is the case if the list and current layout are not compatible
* There can be many lists at once (`awful.layout.layouts`, `t.layouts`, `whatever_the_user_come_up_with`) and they need to be rotated individually.